browser/ui/views/toolbar/wallet_button.cc

```diff
@@ -95,24 +109,71 @@ WalletButton::WalletButton(View* backup_anchor_view, PrefService* prefs)
   SetButtonController(std::move(menu_button_controller));
 
   UpdateVisibility();
+
+  if (brave_wallet::ShouldShowTxStatusInToolbar()) {
+    notification_source_ =
+        std::make_unique<brave::WalletButtonNotificationSource>(
+            profile, base::BindRepeating(&WalletButton::OnNotificationUpdate,
+                                         weak_ptr_factory_.GetWeakPtr()));
+  }
 }
 
 WalletButton::~WalletButton() = default;
 
+void WalletButton::AddedToWidget() {
+  if (notification_source_) {
+    notification_source_->Init();
+  }
+}
+
 void WalletButton::OnWalletPressed(const ui::Event& event) {
   if (IsShowingBubble()) {
     CloseWalletBubble();
     return;
   }
 
   ShowWalletBubble();
+  notification_source_->MarkWalletButtonWasClicked();
+}
+
+void WalletButton::OnNotificationUpdate(bool show_suggest_badge,
+                                        size_t counter) {
+  show_suggest_badge_ = show_suggest_badge;
+  counter_ = counter;
+  UpdateImageAndText();
+}
+
+std::pair<std::string, SkColor> WalletButton::GetBadgeTextAndBackground() {
+  if (counter_ > 0) {
+    std::string text = counter_ > 99 ? "99+" : base::NumberToString(counter_);
+    return {text, brave::kBadgeNotificationBG};
+  }
+  return {"", brave::kBadgeNotificationBG};
 }
 
 void WalletButton::UpdateImageAndText() {
   const ui::ColorProvider* color_provider = GetColorProvider();
   SkColor icon_color = color_provider->GetColor(kColorToolbarButtonIcon);
+  auto icon = gfx::CreateVectorIcon(kWalletToolbarButtonIcon, icon_color);
+
+  size_t icon_size = std::max(icon.width(), icon.height());
+  auto badge_size = brave::BraveIconWithBadgeImageSource::GetBadgeSize();
+  gfx::Size preferred_size(icon_size + badge_size.width() / 2,
+                           icon_size + badge_size.height() / 2);
+
+  auto image_source = std::make_unique<brave::BraveIconWithBadgeImageSource>(
+      preferred_size,
+      base::BindRepeating(&GetColorProviderForView,
+                          weak_ptr_factory_.GetWeakPtr()),
+      icon_size, 5u);
```

Comment:
Where does this `5u` margin value come from?

Reply:
It is used to position the badge on the wallet button. 5u just gives nice picture.

Reply:
Yes I remember now it creates more separation for the badge and the image by shifting the image left. It acts as like a negative left margin. I think the parameter name `content_horizontal_margin` could be confusing as it's not easily understood and takes reading the code to know what it's doing. How about adding a comment as well as naming it something more explicit? I can't think of the perfect name, but perhaps something more like `image_left_margin_extra`? In which case we'd need to change all values currently passed to it to be negative (i.e. `kBraveActionRightMargin * -1` and `-5`). I think this could be worth changing here as this is a new parameter. What do you think?

Reply:
In the shields and rewards buttons, we add insets to the highlight path that so that it still centers the image in the button background (for those buttons it shows on hover) even though the image has been shifted to the left. These are added in `brave_{rewards,shields}_action_view.cc`. Do you not need to add something similar to the wallet button?

Reply:
Fixed